src/coreclr/vm/jithelpers.cpp

```diff
@@ -1793,6 +1803,58 @@ HCIMPL2(void*, JIT_GetSharedNonGCThreadStaticBase, DomainLocalModule *pDomainLoc
     return HCCALL1(JIT_GetNonGCThreadStaticBase_Helper, pMT);
 }
 HCIMPLEND
+
+// *** This helper corresponds CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED.
+//      Even though we always check if the class constructor has been run, we have a separate
+//      helper ID for the "no ctor" version because it allows the JIT to do some reordering that
+//      otherwise wouldn't be possible.
+HCIMPL1(void*, JIT_GetSharedNonGCThreadStaticBaseOptimized, UINT32 staticBlockIndex)
+{
+    void* staticBlock = nullptr;
+
+#ifdef HOST_WINDOWS
+    FCALL_CONTRACT;
+
+    MethodTable * pMT = AppDomain::GetCurrentDomain()->LookupThreadStaticBlockType(staticBlockIndex);
+    _ASSERTE(!pMT->HasGenericsStaticsInfo());
+
+    ENDFORBIDGC();
+    staticBlock = HCCALL1(JIT_GetNonGCThreadStaticBase_Helper, pMT);
+
+    if (t_threadStaticBlocksSize <= staticBlockIndex)
+    {
+        UINT32 prevThreadStaticBlocksSize = t_threadStaticBlocksSize;
+        void** prevThreadStaticBlock = t_threadStaticBlocks;
+
+        t_threadStaticBlocksSize = max(2 * t_threadStaticBlocksSize, staticBlockIndex + 1);
+        t_threadStaticBlocks = (void**) new (nothrow) PTR_BYTE[t_threadStaticBlocksSize * sizeof(PTR_BYTE)];
```

Comment:
This is calling `new (nothrow)` but you are not checking for allocation failure.

This should be calling regular `new` that throws and the whole thing should be under `HELPER_METHOD_FRAME` scope (you can only throw exceptions under `HELPER_METHOD_FRAME` scope).

Reply:
Pure helpers can still throw exceptions like out of memory exception. It is not different from existing thread statics helpers - `JIT_GetNonGCThreadStaticBase_Helper` can throw too.

If the JIT models these helpers as never ever throwing, it is a bug in the JIT.

Reply:
A simplest example is we hoist the helper calls out of loop and if OOM happens, then we are not showing correct information about where the exception happened. That I consider is observable difference in that case that we today take dependency on helpers marked as `pure` . This was the exact scenario I was talking about in https://github.com/dotnet/runtime/pull/82973#issuecomment-1455182469

Reply:
In some cases we might even ignore OOM side effect, e.g. JIT is allowed to remove unused heap allocations